include/range/v3/view/join.hpp

```diff
@@ -58,259 +65,240 @@ namespace ranges
         // Join a range of ranges
         template<typename Rng>
         struct join_view<Rng, void>
-          : view_adaptor<join_view<Rng, void>, Rng,
-                detail::join_cardinality<
-                    range_cardinality<Rng>,
-                    range_cardinality<range_value_t<Rng>>>::value>
+          : view_facade<join_view<Rng, void>, detail::join_cardinality<Rng>::value>
         {
-        private:
-            CONCEPT_ASSERT(Range<Rng>());
-            CONCEPT_ASSERT(Range<range_reference_t<Rng>>());
-            using size_t_ = common_type_t<range_size_t<Rng>, range_size_t<range_value_t<Rng>>>;
+            using size_type = common_type_t<range_size_t<Rng>, range_size_t<range_reference_t<Rng>>>;
 
+            join_view() = default;
+            explicit join_view(Rng rng)
+              : outer_{view::all(std::move(rng))}
+              // Delay initialization of the other members until begin() is first called,
+              // so that we can create a join_view over e.g. an infinite range of empty ranges,
+              // and avoid hanging if begin is never called.
+            {}
+            CONCEPT_REQUIRES(detail::join_cardinality<Rng>::value >= 0)
+            constexpr size_type size() const
+            {
+                return detail::join_cardinality<Rng>::value;
+            }
+            CONCEPT_REQUIRES(detail::join_cardinality<Rng>::value < 0 &&
+                range_cardinality<Rng>::value >= 0 && ForwardRange<Rng>() &&
+                SizedRange<range_reference_t<Rng>>())
+            size_type size() const
+            {
+                return accumulate(view::transform(outer_, ranges::size), size_type{0});
+            }
+        private:
             friend range_access;
-            view::all_t<range_reference_t<Rng>> cur_;
+            Rng outer_{};
 
-            struct adaptor : adaptor_base
+            class cursor : box<range_sentinel_t<Rng>, Rng>
```

Comment:
Using `Rng` as the tag type is likely to make symbols long for no good reason that I can see.